File: AlgorithmsPython/SortingAlgorithms/MergeSort.py
'''
Created on May 3, 2020

@author: Sarthak Jain
'''

'''
Method that sorts an array using MergeSort
'''
def mergeSort(arr, left, right):
    # Goes into this if left is smaller than right
    if left < right:
        
        # Finds the middle between left and right to get the index to divide the array by half
        middle = (left + (right - 1)) // 2
        
        # Recursively goes into the mergeSort method setting left to the left index and middle to the right index 
        mergeSort(arr, left, middle)
        
        # Recursively goes into the mergeSort method setting middle + 1 to the left index and right to the right index 
        mergeSort(arr, middle + 1, right)
        
        # Merges the arrays together 
        merge(arr, left, middle, right)


'''
Merges the items where it sorts an array in a smaller portion of the array
'''
def merge(arr, left, middle, right):
    # Sets the leftSide by subtracting left from one more than middle to get the size of the array
    leftSide = middle - left + 1
    
    # Sets the rightSide by subtracting middle from right to get the size of the array
    rightSide = right - middle
    
    # Creates a LeftArr based on the size of the leftSide
    LeftArr = [0] * (leftSide)

    # Creates a RightArr based on the size of the rightSide
    RightArr = [0] * (rightSide)
    
    # Adds values from the arr to the LeftArr starting from left
    for i in range(0, leftSide):
        LeftArr[i] = arr[left + i]

    # Adds values from the arr to the RightArr starting from middle + 1    
    for i in range(0, rightSide):
        RightArr[i] = arr[middle + 1 + i]
    
    # Initializes variables  
    i = 0
    j = 0
    k = left
    
    # Loop that runs while i is smaller than leftSide and j is smaller than rightSide
    while i < leftSide and j < rightSide:

        # If the value at index i in the leftArr is less than equal to the value at index j in the rightArr, then goes into this condition
        if LeftArr[i] <= RightArr[j]:
        # Sets the value at index k in the arr to the value at index i in the leftArr, and increments i
            arr[k] = LeftArr[i]
            i += 1
        else:
        
        # Otherwise Sets the value at index k in the arr to the value at index j in the rightArr, and increments j
            arr[k] = RightArr[j]
            j += 1
            
        # Increments k by 1
        k += 1
        
    # Loop that runs while i is less than leftSide
    while i < leftSide:

        # Sets the value at index k in the array to the value at index i in the leftArr
        arr[k] = LeftArr[i]
        
        # Increments i and k
        i += 1
        k += 1
    
    # Loop that runs while j is less than rightSide
    while j < rightSide:
        
        # Sets the value at index k in the array to the value at index j in the rightArr
        arr[k] = RightArr[j]
        
        # Increments j and k
        j += 1
        k += 1

'''
Runs the main method where it sorts the array and prints out the sorted array
'''
if __name__ == "__main__": 
    print("Sorts an array using MergeSort")
    arr = [24, 3476984, 11, 22, 69, 0, 88, 34, 331, 1, 23, 732, -1, 4325, 32435432532]
    print("Unsorted Array: " + str(arr));
    mergeSort(arr, 0, len(arr) - 1)
#     mergesort(arr)
    print("Sorted Array: " + str(arr))

File: AlgorithmsJava/src/sortingAlgorithms/MergeSort.java
package sortingAlgorithms;

import java.util.Arrays;
/**
 * A class that implements a generic MergeSort
 * 
 * @author Sarthak Jain
 *
 */
public class MergeSort {
	/**
	 * A private static variable to see when the mergeSort method should switch to
	 * insertion sort
	 */
	private static int switchNum = 3;

	/**
	 * Runs the main method
	 */
	public static void main(String[] args) {
		Integer[] arr = new Integer[] { 24, 3476984, 11, 22, 69, 0, 88, 34, 331, 1, 23, 732, -1, 4325, 324352532 };
		System.out.println("Unsorted Array before MergeSort: " + Arrays.asList(arr));
		mergeSort(arr);
		System.out.println("Sorted Array after MergeSort: " + Arrays.asList(arr));
	}

	/**
	 * The driver method that does the mergesort on an Array from index 0 to the
	 * array length - 1
	 */
	@SuppressWarnings("unchecked")
	public static <T extends Comparable<? super T>> void mergeSort(T[] arr) {

		// Creates a temporary Array with the same size as the userInput Array
		// and sets it all to null
		T[] temp = (T[]) new Comparable[arr.length];
		for (int i = 0; i < arr.length; i++) {
			temp[i] = null;
		}

		// Calls the private merge sort method
		mergeSort(arr, temp, 0, arr.length - 1);
	}

	/**
	 * A private recursive method that splits the Array in half recursively and does
	 * the insertionSort once it meets a certain threshold
	 */
	private static <T extends Comparable<? super T>> void mergeSort(T[] arr, T[] temp, int low, int high) {

		// Once it meets the threshold to switch the condition, then it does the
		// insertion sort
		if (high - low <= switchNum) {
			insertionSort(arr, low, high);
		} else if (low < high) {

			// Finds the middle number
			int mid = (low + high) / 2;

			// Recursively calls this method
			mergeSort(arr, temp, low, mid);
			mergeSort(arr, temp, mid + 1, high);

			// Does the merging
			merge(arr, temp, low, mid + 1, high);
		}

		// Base case
		else {

		}
	}

	/**
	 * A private helper method that does the merging of the two Arrays
	 */
	private static <T extends Comparable<? super T>> void merge(T[] arr, T[] temp, int low, int mid, int high) {

		// Sets the leftIndex of the left half
		int leftArrIndex = low;

		// Sets the rightIndex of the right half
		int rightArrIndex = mid;

		// Sets the index to low
		int index = low;

		// Makes two conditions where it adds the values from the arr to the temporary
		// Array by going from the leftArrIndex to mid -1 and the rightArrIndex to
		// high
		while (leftArrIndex < mid && rightArrIndex <= high) {

			// If the value in the leftArrIndex is less than or equal to the value at the
			// rightArrIndex then it adds it to the temp Array and increments the
			// leftIndex
			if (arr[leftArrIndex].compareTo(arr[rightArrIndex]) <= 0) {
				temp[index] = arr[leftArrIndex];
				leftArrIndex++;

			} else {

				// If the value in the leftArrIndex is greater than to the value at the
				// rightArrIndex then it adds it to the temp Array and increments the
				// rightIndex
				temp[index] = arr[rightArrIndex];
				rightArrIndex++;
			}

			// Increments index
			index++;
		}

		// Another check to make sure that all values are copied over and not missed in
		// the left hand side of the arr and adds it to the temporary Array
		while (leftArrIndex < mid) {
			temp[index] = arr[leftArrIndex];
			leftArrIndex++;
			index++;
		}

		// Another check to make sure that all values are copied over and not missed in
		// the right hand side of the arr and adds it to the temporary Array
		while (rightArrIndex <= high) {
			temp[index] = arr[rightArrIndex];
			rightArrIndex++;
			index++;
		}

		// Copies the values from the temporary array to the original arr
		for (int i = low; i <= high; i++) {
			arr[i] = temp[i];
		}

	}

	/**
	 * A private helper method that does the insertion sort from low to high to sort
	 * the values in order, and is used once high-low is less than equal to the
	 * switchNum threshold in the private merge sort method
	 */
	private static <T extends Comparable<? super T>> void insertionSort(T[] arr, int low, int high) {

		// Creates a for loop that iterates through the array length
		for (int i = low + 1; i <= high; i++) {

			// Creates a temporary value at index at Array
			T val = arr[i];
			int j;

			// Creates another for loop that compares the values of the array at index j and
			// at the temporary value, and compares if the array at index j is greater than
			// the value, then it enters this loop
			for (j = i - 1; j >= low && (arr[j]).compareTo(val) > 0; j--) {

				// Switches the values in the Array
				arr[j + 1] = arr[j];
			}

			arr[j + 1] = val;
		}
	}

}
